quickcheck/jasonkuhrt

jasonkuhrt opened the pull request:
Work in progress.
- [x] Multi-param functions
- [ ] Shrinking
- [ ] Count exceptions as failures too.

jasonkuhrt committed:
quickcheck: Scaffold.

jasonkuhrt committed:
quickcheck: Use Random IO.

jasonkuhrt committed:
quickcheck: Implement `check`.

jasonkuhrt commented:
@neongreen I'm tempted to work on intelligent testcase distribution. Right now its literally just random wrapped up, without any additional logic. However I don't think this is part of the spec right? Meaning what I have so far satisfies the goal.

neongreen commented:
@jasonkuhrt QuickCheck doesn't distribute the cases intelligently but it does try to shrink them intelligently so it still works (e.g. if your function fails on a random string with non-ASCII characters, shrinking will try to bring them to `abcABC123` and find a case that still fails). I guess you can add `weirdness` as a parameter alongside with `size`, but it's not in the spec, right (and not even in QuickCheck).

jasonkuhrt commented:
Makes sense! Thanks for clarifying.

jasonkuhrt committed:
quickcheck: Tidy

jasonkuhrt committed:
quickcheck: First stab at multi-parm functions.

jasonkuhrt committed:
quickcheck: refactor: Monad composition operator.

jasonkuhrt committed:
quickcheck: Tidy.

jasonkuhrt commented:
@neongreen I made some progress on multi-param functions. However I have not been able to unify `Testable` typeclass with `check` (hence the `foobar` function).

`check` created a list of test cases which were inputed into the invariant. Upon a failure (the invariant being broke) the offending test case could be looked up from said list and displayed to the user for them to investigate.

`foobar` is different in that it does _not_ create a list of test cases but rather `test` method from `Testable` typeclass supplies one test case per invocation. In other words a test case is inside `test` therefore hidden from the caller `foobar`. Consequently this prevents `foobar` from telling the user _which_ test cases fail.

I started a solution that revolved around making `test` method return `IO (Maybe [...])` instead of `IO Bool`; The idea being that we could signal failure with `Just ...` including the inputs therein. But of course the inputs may be different types and thus my idea was impossible: Different instances of Testable (functions of various arity) would each need their own different return type.

Another solution I thought of is that `test` method could do the reporting to the user (it is after all in an IO context). This however seems clearly wrong. The separation of concerns becomes totally blurred.

I'll need to think about this more. It seems clear to me that we need to get the `arbitrary` out of `test` somehow. Sleeping on this. Any hints appreciated.

neongreen commented:
That's a really good question, because I don't know the answer :)

[after some source-reading]

QuickCheck goes the “test method does the reporting” route. Specifically, `Property` is a thing that randomly constructs a complicated structure that contains callbacks that should be called after the test case. As the result, it's actually impossible to get the failing testcase out of QuickCheck – it can only be printed.

You can go a different route:

``` haskell
{-# LANGUAGE TypeFamilies #-}

class Testable a where
  type TestCase a
  test :: a -> IO (Maybe (TestCase a))

instance Show a => Testable (a -> Bool) where
  type TestCase (a -> Bool) = a
  test = ...

instance (Show a, Show b) => Testable (a -> b -> Bool) where
  type TestCase (a -> b -> Bool) = (a, b)
  test = ...
```

jasonkuhrt commented:
@neongreen Thanks so much. I will try working with type families as it seems to provide more freedom and expressivity.

jasonkuhrt committed:
quickcheck: Allow `test` method to return failed test cases.

Key changes:

* Use extension TypeFamilies
* Use extension FlexibleContexts to permit classifying `Testcase a`
  so that the result of `test` method can be e.g. printed.

jasonkuhrt committed:
quickcheck: Expose all test cases.

`test` method used to only expose test cases that invalidated the
property being tested. However for certain modes (e.g. verbose) we
want to show all test cases, fail or pass. To achieve this we replace
Maybe with Either. Test cases that invalidate the property are in the
Left constructor, test cases that pass are in Right.

jasonkuhrt committed:
quickcheck: Finish new `check` function

Rename the previously WIP `foobar` function to `check` and remove the
old `check` implementation.

jasonkuhrt commented:
Ok, I think I've finished the multi-param functions task, woo-hoo.
